File: BaekJoon/동적 계획법 1/9251_LCS.cpp
#include <iostream>
#include <string>
using namespace std;

int max(int a, int b){
    return a > b ? a : b;
}

int main(){
	string a, b;
	int row, column;
	int dp[1001][1001] = {0,};
	
	cin >> a >> b;
	
	row = a.length();
	column = b.length();
	
	for(int i = 1; i <= row; i++){
	    for(int j = 1; j <= column; j++){
	        if(a[i-1] == b[j-1]){
	            dp[i][j] = dp[i-1][j-1] + 1;
	        }
	        else{
	            dp[i][j] = max(dp[i-1][j], dp[i][j-1]);
	        }
	    }
	}
	
	cout << dp[row][column];
}

File: BaekJoon/동적 계획법 1/9251_LCS.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.lang.Integer.max

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val first = readLine().chunked(1)
    val second = readLine().chunked(1)

    val row = first.size
    val col = second.size

    val dp = Array(row + 1) { Array(col + 1) { 0 } }

    for (i in 1..row) {
        for (j in 1..col) {
            if (first[i - 1] == second[j - 1]) {
                dp[i][j] = dp[i - 1][j - 1] + 1
            } else {
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
            }
        }
    }

    println(dp[row][col])
}
